src/app/util/ember-compatibility-functions.cpp

```diff
@@ -340,5 +344,121 @@ CHIP_ERROR ReadSingleClusterData(ClusterInfo & aClusterInfo, TLV::TLVWriter * ap
     return CHIP_NO_ERROR;
 }
 
+namespace {
+template <typename T>
+CHIP_ERROR numericTlvDataToAttributeBuffer(TLV::TLVReader & aReader, uint16_t & dataLen)
+{
+    T value;
+    static_assert(sizeof(value) <= sizeof(attributeData), "Value cannot fit into attribute data");
+    dataLen = sizeof(value);
+    ReturnErrorOnFailure(aReader.Get(value));
+    memcpy(attributeData, &value, sizeof(value));
+    return CHIP_NO_ERROR;
+}
+template <typename T>
+CHIP_ERROR stringTlvDataToAttributeBuffer(TLV::TLVReader & aReader, uint16_t & dataLen)
+{
+    const uint8_t * data = nullptr;
+    T len;
+    VerifyOrReturnError(aReader.GetType() == TLV::TLVType::kTLVType_ByteString ||
+                            aReader.GetType() == TLV::TLVType::kTLVType_UTF8String,
+                        CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(CanCastTo<T>(aReader.GetLength()), CHIP_ERROR_MESSAGE_TOO_LONG);
+    ReturnErrorOnFailure(aReader.GetDataPtr(data));
+    len = static_cast<T>(aReader.GetLength());
+    VerifyOrReturnError(len + sizeof(len) /* 1 byte of length */ <= sizeof(attributeData), CHIP_ERROR_MESSAGE_TOO_LONG);
```

Comment:
Where does the `1 byte of length` comes from ? My understanding is that `sizeof(T)` here represents either `uint8_t` or `uint16_t` ?